builder/builder_test.go

```diff
@@ -202,16 +215,152 @@ func testBuilder(t *testing.T, when spec.G, it spec.S) {
+						h.AssertNil(t, json.Unmarshal([]byte(label), &metadata))
+
+						h.AssertEq(t, metadata.Groups[0].Buildpacks[0].ID, "buildpack-1-id")
+						h.AssertEq(t, metadata.Groups[0].Buildpacks[0].Version, "buildpack-1-version")
+					})
+				})
+
+				when("has multiple buildpacks with same ID", func() {
+					it.Before(func() {
+						subject.AddBuildpack(buildpack.Buildpack{
+							BuildpackInfo: buildpack.BuildpackInfo{
+								ID:      "buildpack-1-id",
+								Version: "buildpack-1-version",
+							},
+							Path: buildpackTgz,
+							Stacks: []buildpack.Stack{
+								{ID: "some.stack.id"},
+							},
+						})
+
+						subject.AddBuildpack(buildpack.Buildpack{
+							BuildpackInfo: buildpack.BuildpackInfo{
+								ID:      "buildpack-1-id",
+								Version: "buildpack-1-version2",
+							},
+							Path: buildpackTgz,
+							Stacks: []buildpack.Stack{
+								{ID: "some.stack.id"},
+							},
+						})
+					})
+
+					when("order explicitly sets version", func() {
+						it("should keep order version", func() {
+							subject.SetOrder(builder.Order{{
+								Group: []builder.BuildpackRef{
+									{BuildpackInfo: buildpack.BuildpackInfo{ID: "buildpack-1-id", Version: "buildpack-1-version"}}},
+							}})
+
+							err := subject.Save()
+							h.AssertNil(t, err)
+
+							label, err := baseImage.Label("io.buildpacks.builder.metadata")
+							h.AssertNil(t, err)
+
+							var metadata builder.Metadata
+							h.AssertNil(t, json.Unmarshal([]byte(label), &metadata))
+
+							// adds new buildpack
+							h.AssertEq(t, metadata.Groups[0].Buildpacks[0].ID, "buildpack-1-id")
+							h.AssertEq(t, metadata.Groups[0].Buildpacks[0].Version, "buildpack-1-version")
+						})
+					})
+
+					when("order version is empty", func() {
+						it("return error", func() {
+							subject.SetOrder(builder.Order{{
+								Group: []builder.BuildpackRef{
+									{BuildpackInfo: buildpack.BuildpackInfo{ID: "buildpack-1-id"}}},
+							}})
+
+							err := subject.Save()
+							h.AssertError(t, err, "multiple versions of 'buildpack-1-id' - must specify an explicit version")
+						})
+					})
+				})
+
+				when("order points to missing buildpack id", func() {
+					it.Before(func() {
+						subject.AddBuildpack(buildpack.Buildpack{
+							BuildpackInfo: buildpack.BuildpackInfo{
+								ID:      "buildpack-1-id",
+								Version: "buildpack-1-version",
+							},
+							Path: buildpackTgz,
+							Stacks: []buildpack.Stack{
+								{ID: "some.stack.id"},
+							},
+						})
+					})
+
+					it("should error", func() {
+						subject.SetOrder(builder.Order{{
+							Group: []builder.BuildpackRef{
+								{BuildpackInfo: buildpack.BuildpackInfo{ID: "missing-buildpack-id"}}},
+						}})
+
+						err := subject.Save()
+
+						h.AssertError(t, err, "no versions of buildpack 'missing-buildpack-id' were found on the builder")
+					})
+				})
+
+				when("order points to missing buildpack version", func() {
```

Comment:
Same comment as above, we could skip respecifying the same Before steps if we move the its to the other `when` block.